### ISSUE TITLE
Navigation strategy should support up navigation

### ISSUE THREAD

**Issue by rala72** [consp1racy/android-support-preference#116]
I'm using the navigation between subscreens like the example shows...
I can also call a subscreen directly - but after navigating back the app closes.

**Comment by consp1racy** [comment]
Look, if this is all information you can give then I can't help you. I don't know what Android this happened on, I don't know the precise error, I don't know the exact code which you used or the dependencies, I don't know anything at all. I want to help you but the best I can do now is guess and that's just not good enough.

Make a minimal sample project that demonstrates the issue, put it on Github and post a link to it here. Then I may be able to help.

Thanks, @rala72

**Comment by rala72** [comment]
I'm really sorry about that.

- Android version _(I used for testing)_: 9.0
- Device manufacturer and model: OP5
- Stacktrace with line numbers preserved: _no stack trace_
- Android support library version: 1.0.+
- xpece-android-support-preference library version: 2.3.+

the code _(for subscreens)_ is nearly 100% the same as the example shows.

I adapted my code to add a flag if it is called directly - and if yes it finds the parent screen in on `onBackPressed` in the activity and load the fragment there.
Without adaption the app gets closed with `super.onBackPressed();`

```java
public static PreferenceFragmentCompat newInstance(String rootKey, boolean directCalled) {
    PreferenceFragment fragment = new PreferenceFragment();
    Bundle bundle = new Bundle();
    bundle.putString(PreferenceFragment.ARG_PREFERENCE_ROOT, rootKey);
    bundle.putBoolean(PreferenceFragment.ARG_DIRECT_CALLED, directCalled);
    fragment.setArguments(bundle);
    return fragment;
}
```

**Comment by consp1racy** [comment]
That's loads better, thanks.

So you're able to open the settings activity and navigate the whole preference screen hierarchy in the correct order.

And in another part of the app you allow opening a subscreen of settings and you want the up navigation button to lead to settings root before closing the settings activity?

I'm sorry if I don't understand correctly. If I got it wrong please draw a diagram.

**Comment by rala72** [comment]
no you are getting me correct.
in my case the settings root - but it would be nice if it works always (meaning even if you have more than 2 layers)

**Comment by consp1racy** [comment]
Ok, I've got two solutions for you.

## The ugly, wrong, and incomplete

The following snippet belongs in the activity. It's a slightly modified bit of the example `SettingsActivity`.

    mReplaceFragmentStrategy = new PreferenceScreenNavigationStrategy.ReplaceFragment(this, R.anim.abc_fade_in, R.anim.abc_fade_out, R.anim.abc_fade_in, R.anim.abc_fade_out);

    if (savedInstanceState == null) {
        final String key = getIntent().getStringExtra(SettingsFragment.ARG_PREFERENCE_ROOT);
        final Fragment settingsFragment = onBuildPreferenceFragment(key);
        final FragmentManager fm = getSupportFragmentManager();
        if (key == null) {
            fm.beginTransaction()
                    .add(R.id.content, settingsFragment, "Settings")
                    .commitNow();
        } else {
            final Fragment rootSettingsFragment = onBuildPreferenceFragment(null);
            fm.beginTransaction()
                    .add(R.id.content, rootSettingsFragment, "Settings")
                    .commit();
            fm.beginTransaction()
                    .setCustomAnimations(R.anim.abc_fade_in, R.anim.abc_fade_out, R.anim.abc_fade_in, R.anim.abc_fade_out)
                    .addToBackStack(null)
                    .replace(R.id.content, settingsFragment, "Settings")
                    .commit();
            fm.executePendingTransactions();
        }
    }

### Behavior

When we get the preference key from an intent, we put a root fragment at the bottom and then put the target fragment on top and in the back stack.

When you press the Back button, you go to root preference screen. This is bad from the point of Android navigation guidelines. The system Back button should return you where you came from.

The Up button in this case delegates to the Back button. This is also wrong because the Up button, per guidelines, should navigate you through the whole stack in the current app. In this case it should navigate through each intermediary preference screen to the root, and only then leave the activity (and potentially open hierarchical parent).

### Implementation

Note the single `commitNow()` versus multiple `commit()` + `executePendingTransactons()` calls. You also have to manually call `setCustomAnimations()` with the same values you supplied to the replace strategy constructor.

I'm trying to reuse as much code as possible, that's why I call `onBuildPreferenceFragment()` - it already knows how to create a preference fragment for a key.

## The correct

Here's a more "correct" approach that only changes the Up button behavior.

The part that goes into `Activity.onCreate`:

    if (savedInstanceState == null) {
        final String key = getIntent().getStringExtra(SettingsFragment.ARG_PREFERENCE_ROOT);
        final Fragment settingsFragment = onBuildPreferenceFragment(key);
        getSupportFragmentManager().beginTransaction()
                .add(R.id.content, settingsFragment, "Settings")
                .commitNow();
    }

Handling of the Up button (in activity):

    @Override
    public boolean onOptionsItemSelected(@NonNull final MenuItem item) {
        int id = item.getItemId();
        switch (id) {
            case android.R.id.home: {
                final FragmentManager fm = getSupportFragmentManager();
                final SettingsFragment f = (SettingsFragment) fm.findFragmentByTag("Settings");
                final PreferenceScreen screen = f.getPreferenceScreen();
                final PreferenceGroup parent = screen.getParent();
                if (parent == null || fm.getBackStackEntryCount() != 0) {
                    // If we're at root or we have a back stack let the activity do its thing.
                    onBackPressed();
                } else {
                    // Otherwise synthesize the back stack on-the-fly.
                    final String key = parent.getKey();
                    final Fragment newFragment = onBuildPreferenceFragment(key);
                    fm.beginTransaction()
                            .setCustomAnimations(R.anim.abc_fade_in, R.anim.abc_fade_out, R.anim.abc_fade_in, R.anim.abc_fade_out)
                            .replace(f.getId(), newFragment, f.getTag())
                            .commitNow();
                }
                return true;
            }
        return false;
    }

Fix the root preference activity title (in your preference fragment):

    @Override
    public void onCreatePreferences2(final Bundle savedInstanceState, final String rootKey) {
        // ...

        // Setup root preference title.
        final PreferenceScreen root = getPreferenceScreen();
        if (TextUtils.isEmpty(root.getTitle())) {
            // Only set title *of the root preference* if it's empty.
            // getActivity().getTitle() doesn't work correctly with synthesized back stack.
            // final FragmentActivity activity = getActivity();
            // final PackageManager pm = activity.getPackageManager();
            // final CharSequence title = pm.getActivityInfo(activity.getComponentName(), 0).loadLabel(pm);
            // final CharSequence title = activity.getApplicationInfo().loadLabel(pm);
            final CharSequence title = getString(R.string.settings_activity_title);
            root.setTitle(title);
        }

        // Setup root preference.
        // Use with ReplaceFragment strategy.
        PreferenceScreenNavigationStrategy.ReplaceFragment.onCreatePreferences(this, rootKey);
    }

### Behavior

The Back button will leave the activity to where you came from. The Up button will open parent preference screens until you're at root and then it leaves the activity.

### Implementation

Only the target preference fragment is attached at start. Parent preference fragments are created and replaced as the user navigates up.

You still have to specify the animations manually.

The parent fragments are created at the moment of navigation. The activity title is already changed by then and not representative of root preference. That's why we can't use `getActivity().getTitle()`. Several alternative options are listed above with the fastest uncommented.

**Comment by consp1racy** [comment]
I think I could expand the navigation strategy class to support Up navigation. This is a good idea, thanks!

**Comment by consp1racy** [comment]
Will be released in 2.4.0.

Usage:

```java
@Override
public boolean onOptionsItemSelected(@NonNull final MenuItem item) {
    final int id = item.getItemId();
    switch (id) {
        case android.R.id.home: {
            final FragmentManager fm = getSupportFragmentManager();
            final SettingsFragment f = (SettingsFragment) fm.findFragmentByTag("Settings");
            if (!mReplaceFragmentStrategy.onNavigateUp(fm, f)) {
                // If we're at root let the activity do its thing.
                onBackPressed();
            }
            return true;
        }
    }
    return false;
}
```
